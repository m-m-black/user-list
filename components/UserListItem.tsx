import { View, StyleSheet, useColorScheme } from "react-native";
import { ListItem, Avatar } from "@rneui/themed";

import {
  Colors,
  avatarBackground,
  avatarText,
  avatarSubtitleText,
} from "@/constants/Colors";

type UserListItemProps = {
  name: string;
  role: string;
};

export default function UserListItem({ name, role }: UserListItemProps) {
  const theme = useColorScheme() ?? "light";

  const avatarLabel = name[0].toUpperCase();

  // Custom styles based on device theme (light or dark)
  const themedStyles = {
    container: {
      backgroundColor:
        theme === "light" ? Colors.light.background : Colors.dark.background,
    },
    titleText: {
      color: theme === "light" ? Colors.light.text : Colors.dark.text,
    },
  };

  return (
    <View>
      <ListItem containerStyle={themedStyles.container}>
        <Avatar
          size={40}
          title={avatarLabel}
          containerStyle={styles.avatarContainer}
          titleStyle={styles.avatarText}
        />
        <ListItem.Content>
          <ListItem.Title style={themedStyles.titleText}>{name}</ListItem.Title>
          <ListItem.Subtitle style={styles.subtitleText}>
            {role}
          </ListItem.Subtitle>
        </ListItem.Content>
      </ListItem>
    </View>
  );
}

const styles = StyleSheet.create({
  avatarContainer: {
    backgroundColor: avatarBackground,
    borderRadius: 5,
  },
  avatarText: {
    color: avatarText,
  },
  subtitleText: {
    color: avatarSubtitleText,
  },
});
