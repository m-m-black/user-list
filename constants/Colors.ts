const tintColorLight = "#306FC7";
const tintColorDark = "#fff";

// These colours remain consistent across themes
export const avatarBackground = "#EAF2FA";
export const avatarText = "#306FC7";
export const avatarSubtitleText = "#808080";

export const Colors = {
  light: {
    text: "#11181C",
    background: "#fff",
    backgroundSecondary: "#EAF2FA",
    tint: tintColorLight,
    icon: "#687076",
    tabIconDefault: "#687076",
    tabIconSelected: tintColorLight,
  },
  dark: {
    text: "#ECEDEE",
    background: "#151718",
    backgroundSecondary: "#2F4F4F",
    tint: tintColorDark,
    icon: "#9BA1A6",
    tabIconDefault: "#9BA1A6",
    tabIconSelected: tintColorDark,
  },
};
