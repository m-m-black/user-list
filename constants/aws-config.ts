// These would normally be defined in a .env file
export const AWS_APPSYNC_URL =
  "https://prrwjjssnvhpbcdwbcwx3nm3zm.appsync-api.ap-southeast-2.amazonaws.com/graphql";
export const AWS_APPSYNC_API_KEY = "da2-6y6arb7mwvgrnmds2jignrgr2u";
