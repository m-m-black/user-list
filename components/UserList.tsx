import { View, Text, StyleSheet, FlatList } from "react-native";
import { gql, useQuery } from "@apollo/client";

import UserListItem from "./UserListItem";

import { UserRole } from "@/constants/UserRole";

// TODO: Delete this
const testCustomers = [
  {
    id: "1",
    name: "TestCustomer1",
    email: "test1@test.com",
    role: "Manager",
  },
  {
    id: "2",
    name: "TestCustomer2",
    email: "test2@test.com",
    role: "Admin",
  },
  {
    id: "3",
    name: "TestCustomer3",
    email: "test3@test.com",
    role: "Manager",
  },
  {
    id: "4",
    name: "TestCustomer4",
    email: "test4@test.com",
    role: "Admin",
  },
];

// Export this query to make it available it tests
export const LIST_ZELLER_CUSTOMERS = gql`
  query ListZellerCustomers(
    $filter: TableZellerCustomerFilterInput
    $limit: Int
    $nextToken: String
  ) {
    listZellerCustomers(filter: $filter, limit: $limit, nextToken: $nextToken) {
      items {
        id
        name
        email
        role
      }
      nextToken
    }
  }
`;

type UserListProps = {
  selectedRole: UserRole | undefined;
};

export default function UserList({ selectedRole }: UserListProps) {
  // Set query filter based on selectedRole
  const roleFilter = selectedRole ? { role: { eq: selectedRole } } : null;

  // Define query for data fetching (and re-fetching) with filter
  const { loading, error, data } = useQuery(LIST_ZELLER_CUSTOMERS, {
    variables: { filter: roleFilter, limit: 10 },
  });

  if (loading) {
    return (
      <View style={styles.container}>
        <View style={styles.statusContainer}>
          <Text>Loading...</Text>
        </View>
      </View>
    );
  }

  if (error) {
    return (
      <View style={styles.container}>
        <View style={styles.statusContainer}>
          <Text>Error: {error.message}</Text>
        </View>
      </View>
    );
  }

  return (
    <View style={styles.container}>
      <FlatList
        // data={testCustomers} // TODO: Delete this
        data={data.listZellerCustomers.items}
        contentContainerStyle={styles.listContainer}
        renderItem={({ item, index }) => (
          <UserListItem name={item.name} role={item.role} key={index} />
        )}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
  },
  statusContainer: {
    alignItems: "center",
    justifyContent: "center",
  },
  listContainer: {
    flex: 1,
  },
});
