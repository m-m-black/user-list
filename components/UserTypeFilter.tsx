import { View, StyleSheet } from "react-native";
import { CheckBox } from "@rneui/themed";

import { UserRole } from "@/constants/UserRole";

type UserTypeFilterProps = {
  selectedRole: UserRole | undefined;
  setSelectedRole: (role: UserRole | undefined) => void;
};

export default function UserTypeFilter({
  selectedRole,
  setSelectedRole,
}: UserTypeFilterProps) {
  const toggleFilter = (role: UserRole) => {
    if (selectedRole === role) {
      setSelectedRole(undefined);
    } else {
      setSelectedRole(role);
    }
  };

  return (
    <View>
      <CheckBox
        checked={selectedRole === UserRole.ADMIN}
        onPress={() => toggleFilter(UserRole.ADMIN)}
        checkedIcon="dot-circle-o"
        uncheckedIcon="circle-o"
        title="Admin"
        containerStyle={[
          styles.container,
          selectedRole === UserRole.ADMIN && styles.containerSelected,
        ]}
        checkedColor="#306FC7"
      />
      <CheckBox
        checked={selectedRole === UserRole.MANAGER}
        onPress={() => toggleFilter(UserRole.MANAGER)}
        checkedIcon="dot-circle-o"
        uncheckedIcon="circle-o"
        title="Manager"
        containerStyle={[
          styles.container,
          selectedRole === UserRole.MANAGER && styles.containerSelected,
        ]}
        checkedColor="#306FC7"
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    borderRadius: 10,
  },
  containerSelected: {
    backgroundColor: "#EAF2FA",
  },
});
