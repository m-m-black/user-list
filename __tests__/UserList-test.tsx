import React, { useState } from "react";
import { render, fireEvent, waitFor, act } from "@testing-library/react-native";
import { MockedProvider } from "@apollo/client/testing";

import UserTypeFilter from "@/components/UserTypeFilter";
import UserList from "@/components/UserList";
import { LIST_ZELLER_CUSTOMERS } from "@/components/UserList";

import { UserRole } from "@/constants/UserRole";

const testCustomers = [
  {
    id: "1",
    name: "TestCustomer1",
    email: "test1@test.com",
    role: "Manager",
  },
  {
    id: "2",
    name: "TestCustomer2",
    email: "test2@test.com",
    role: "Admin",
  },
  {
    id: "3",
    name: "TestCustomer3",
    email: "test3@test.com",
    role: "Manager",
  },
  {
    id: "4",
    name: "TestCustomer4",
    email: "test4@test.com",
    role: "Admin",
  },
];

// Wrapper component to handle selectedRole state
const TestWrapper = ({
  initialRole = undefined,
}: {
  initialRole: UserRole | undefined;
}) => {
  const [selectedRole, setSelectedRole] = useState<UserRole | undefined>(
    initialRole
  );

  return (
    <>
      <UserTypeFilter
        selectedRole={selectedRole}
        setSelectedRole={setSelectedRole}
      />
      <UserList selectedRole={selectedRole} />
    </>
  );
};

// All components used from React Native Elements are mocked.
// See __mocks__/@rneui/themed.js
jest.mock("@rneui/themed");

describe("UserList", () => {
  const mocks = [
    // Mock for all customers (no filter applied)
    {
      request: {
        query: LIST_ZELLER_CUSTOMERS,
        variables: { filter: null, limit: 10 },
      },
      result: {
        data: {
          listZellerCustomers: {
            items: testCustomers,
            nextToken: null,
          },
        },
      },
    },
    // Mock for Admin filter
    {
      request: {
        query: LIST_ZELLER_CUSTOMERS,
        variables: {
          filter: { role: { eq: UserRole.ADMIN } },
          limit: 10,
        },
      },
      result: {
        data: {
          listZellerCustomers: {
            items: testCustomers.filter((c) => c.role === "Admin"),
            nextToken: null,
          },
        },
      },
    },
    // Mock for Manager filter
    {
      request: {
        query: LIST_ZELLER_CUSTOMERS,
        variables: {
          filter: { role: { eq: UserRole.MANAGER } },
          limit: 10,
        },
      },
      result: {
        data: {
          listZellerCustomers: {
            items: testCustomers.filter((c) => c.role === "Manager"),
            nextToken: null,
          },
        },
      },
    },
  ];

  test("shows all users initially and filters to Admin users when Admin is selected", async () => {
    const { getByText, queryByText } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <TestWrapper initialRole={undefined} />
      </MockedProvider>
    );

    // Initially should show loading state
    expect(getByText("Loading...")).toBeTruthy();

    // Wait for initial data to load
    await waitFor(() => {
      expect(queryByText("Loading...")).toBeNull();
    });

    // Should show all users initially
    expect(getByText("TestCustomer1")).toBeTruthy();
    expect(getByText("TestCustomer2")).toBeTruthy();
    expect(getByText("TestCustomer3")).toBeTruthy();
    expect(getByText("TestCustomer4")).toBeTruthy();

    // Click Admin radio button
    const adminButton = getByText("◯ Admin");

    act(() => {
      fireEvent.press(adminButton);
    });

    // Should show loading state again
    expect(getByText("Loading...")).toBeTruthy();

    // Wait for filtered data to load
    await waitFor(() => {
      expect(queryByText("Loading...")).toBeNull();
    });

    // Should only show Admin users
    expect(queryByText("TestCustomer1")).toBeNull(); // Manager
    expect(getByText("TestCustomer2")).toBeTruthy(); // Admin
    expect(queryByText("TestCustomer3")).toBeNull(); // Manager
    expect(getByText("TestCustomer4")).toBeTruthy(); // Admin
  });

  test("shows only Manager users when Manager is selected", async () => {
    const { getByText, queryByText } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <TestWrapper initialRole={undefined} />
      </MockedProvider>
    );

    await waitFor(() => {
      expect(queryByText("Loading...")).toBeNull();
    });

    // Click Manager radio button
    const managerButton = getByText("◯ Manager");

    act(() => {
      fireEvent.press(managerButton);
    });

    // Wait for filtered data to load
    await waitFor(() => {
      expect(queryByText("Loading...")).toBeNull();
    });

    // Should only show Manager users
    expect(getByText("TestCustomer1")).toBeTruthy(); // Manager
    expect(queryByText("TestCustomer2")).toBeNull(); // Admin
    expect(getByText("TestCustomer3")).toBeTruthy(); // Manager
    expect(queryByText("TestCustomer4")).toBeNull(); // Admin
  });

  test("shows all users again when selected role is cleared", async () => {
    const { getByText, queryByText } = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <TestWrapper initialRole={UserRole.ADMIN} />
      </MockedProvider>
    );

    await waitFor(() => {
      expect(queryByText("Loading...")).toBeNull();
    });

    // Initially should only show Admin users
    expect(queryByText("TestCustomer1")).toBeNull(); // Manager
    expect(getByText("TestCustomer2")).toBeTruthy(); // Admin
    expect(queryByText("TestCustomer3")).toBeNull(); // Manager
    expect(getByText("TestCustomer4")).toBeTruthy(); // Admin

    // Click Admin radio button again to clear
    const adminButton = getByText("⚫ Admin");

    act(() => {
      fireEvent.press(adminButton);
    });

    // Wait for unfiltered data to load
    await waitFor(() => {
      expect(queryByText("Loading...")).toBeNull();
    });

    // Should show all users
    expect(getByText("TestCustomer1")).toBeTruthy();
    expect(getByText("TestCustomer2")).toBeTruthy();
    expect(getByText("TestCustomer3")).toBeTruthy();
    expect(getByText("TestCustomer4")).toBeTruthy();
  });

  // Snapshot test to ensure UI remains consistent
  test("it renders correctly", () => {
    const tree = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <TestWrapper initialRole={UserRole.ADMIN} />
      </MockedProvider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
