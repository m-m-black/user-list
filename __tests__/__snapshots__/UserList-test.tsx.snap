// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList it renders correctly 1`] = `
[
  <View>
    <View
      checked={true}
      checkedColor="#306FC7"
      checkedIcon="dot-circle-o"
      containerStyle={
        [
          {
            "borderRadius": 10,
          },
          {
            "backgroundColor": "#EAF2FA",
          },
        ]
      }
      onPress={[Function]}
      title="Admin"
      uncheckedIcon="circle-o"
    >
      <Text>
        ⚫
         
        Admin
      </Text>
    </View>
    <View
      checked={false}
      checkedColor="#306FC7"
      checkedIcon="dot-circle-o"
      containerStyle={
        [
          {
            "borderRadius": 10,
          },
          false,
        ]
      }
      onPress={[Function]}
      title="Manager"
      uncheckedIcon="circle-o"
    >
      <Text>
        ◯
         
        Manager
      </Text>
    </View>
  </View>,
  <View
    style={
      {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text>
        Loading...
      </Text>
    </View>
  </View>,
]
`;
